File: frontend/src/services/diagnostics.ts
/**
 * Diagnostics service for admin troubleshooting (GH Issue #23)
 */
import { api } from './api';
import {
  AgentDebugStatus,
  AgentDebugStatusResponse,
  AgentLogsResponse,
  DebugToggleResponse,
  LogPurgeResponse,
  SystemInfoResponse,
  ToggleDebugRequest,
  ServerDebugStatus,
  AllLogStats,
  PurgeLogsResponse,
  PostgresLogsExistResponse,
  NginxLogsExistResponse,
  NginxReloadResponse
} from '../types/diagnostics';

// Get system diagnostic info
export const getSystemInfo = async (): Promise<SystemInfoResponse> => {
  const response = await api.get<SystemInfoResponse>('/api/admin/diagnostics/system-info');
  return response.data;
};

// Get debug status for all agents
export const getAgentDebugStatuses = async (): Promise<AgentDebugStatusResponse> => {
  const response = await api.get<AgentDebugStatusResponse>('/api/admin/diagnostics/agents');
  return response.data;
};

// Get debug status for a specific agent
export const getAgentDebugStatus = async (agentId: number): Promise<AgentDebugStatus> => {
  const response = await api.get<AgentDebugStatus>(`/api/admin/diagnostics/agents/${agentId}`);
  return response.data;
};

// Toggle debug mode for a specific agent
export const toggleAgentDebug = async (agentId: number, enable: boolean): Promise<DebugToggleResponse> => {
  const response = await api.post<DebugToggleResponse>(
    `/api/admin/diagnostics/agents/${agentId}/debug`,
    { enable } as ToggleDebugRequest
  );
  return response.data;
};

// Toggle debug mode for all agents
export const toggleAllAgentsDebug = async (enable: boolean): Promise<DebugToggleResponse> => {
  const response = await api.post<DebugToggleResponse>(
    '/api/admin/diagnostics/agents/debug',
    { enable } as ToggleDebugRequest
  );
  return response.data;
};

// Request logs from a specific agent
export const requestAgentLogs = async (
  agentId: number,
  hoursBack: number = 24,
  includeAll: boolean = false
): Promise<AgentLogsResponse> => {
  const params = new URLSearchParams();
  if (hoursBack) params.append('hours_back', hoursBack.toString());
  if (includeAll) params.append('include_all', 'true');

  const response = await api.get<AgentLogsResponse>(
    `/api/admin/diagnostics/agents/${agentId}/logs?${params.toString()}`
  );
  return response.data;
};

// Purge logs for a specific agent
export const purgeAgentLogs = async (agentId: number): Promise<LogPurgeResponse> => {
  const response = await api.delete<LogPurgeResponse>(`/api/admin/diagnostics/agents/${agentId}/logs`);
  return response.data;
};

// Download full diagnostic package
export const downloadDiagnostics = async (
  includeAgentLogs: boolean = false,
  hoursBack: number = 1,
  includeNginxLogs: boolean = false,
  includePostgresLogs: boolean = false
): Promise<Blob> => {
  const params = new URLSearchParams();
  params.append('hours_back', hoursBack.toString());
  if (includeAgentLogs) {
    params.append('include_agent_logs', 'true');
  }
  if (includeNginxLogs) {
    params.append('include_nginx_logs', 'true');
  }
  if (includePostgresLogs) {
    params.append('include_postgres_logs', 'true');
  }
  const response = await api.get(`/api/admin/diagnostics/download?${params.toString()}`, {
    responseType: 'blob'
  });
  return response.data;
};

// Helper function to trigger download of the diagnostic package
export const downloadDiagnosticsFile = async (
  includeAgentLogs: boolean = false,
  hoursBack: number = 1,
  includeNginxLogs: boolean = false,
  includePostgresLogs: boolean = false
): Promise<void> => {
  const blob = await downloadDiagnostics(includeAgentLogs, hoursBack, includeNginxLogs, includePostgresLogs);
  const url = window.URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
  link.download = `krakenhashes-diagnostics-${timestamp}.zip`;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  window.URL.revokeObjectURL(url);
};

// Get server/backend debug status
export const getServerDebugStatus = async (): Promise<ServerDebugStatus> => {
  const response = await api.get<ServerDebugStatus>('/api/admin/diagnostics/server/debug');
  return response.data;
};

// Toggle server/backend debug mode
export const toggleServerDebug = async (enable: boolean, level?: string): Promise<ServerDebugStatus> => {
  const response = await api.post<ServerDebugStatus>(
    '/api/admin/diagnostics/server/debug',
    { enable, level } as ToggleDebugRequest
  );
  return response.data;
};

// Get server log stats
export const getLogStats = async (): Promise<AllLogStats> => {
  const response = await api.get<AllLogStats>('/api/admin/diagnostics/logs/stats');
  return response.data;
};

// Purge server logs
export const purgeServerLogs = async (directory: 'backend' | 'nginx' | 'postgres' | 'all'): Promise<PurgeLogsResponse> => {
  const response = await api.delete<PurgeLogsResponse>(`/api/admin/diagnostics/logs/${directory}`);
  return response.data;
};

// Check if PostgreSQL logs exist
export const checkPostgresLogsExist = async (): Promise<PostgresLogsExistResponse> => {
  const response = await api.get<PostgresLogsExistResponse>('/api/admin/diagnostics/postgres-logs-exist');
  return response.data;
};

// Check if Nginx logs exist
export const checkNginxLogsExist = async (): Promise<NginxLogsExistResponse> => {
  const response = await api.get<NginxLogsExistResponse>('/api/admin/diagnostics/nginx-logs-exist');
  return response.data;
};

// Reload nginx configuration (hot-reload)
export const reloadNginx = async (): Promise<NginxReloadResponse> => {
  const response = await api.post<NginxReloadResponse>('/api/admin/diagnostics/nginx/reload');
  return response.data;
};

File: backend/internal/routes/diagnostics.go
package routes

import (
	"database/sql"
	"net/http"

	"github.com/ZerkerEOD/krakenhashes/backend/internal/db"
	admindiagnostics "github.com/ZerkerEOD/krakenhashes/backend/internal/handlers/admin/diagnostics"
	wshandler "github.com/ZerkerEOD/krakenhashes/backend/internal/handlers/websocket"
	"github.com/ZerkerEOD/krakenhashes/backend/internal/middleware"
	"github.com/ZerkerEOD/krakenhashes/backend/internal/services/diagnostic"
	"github.com/ZerkerEOD/krakenhashes/backend/pkg/debug"
	"github.com/gorilla/mux"
)

// SetupDiagnosticsRoutes configures the admin diagnostics routes (GH Issue #23)
// This is called after WebSocket handler is initialized since it requires access to agent connections.
func SetupDiagnosticsRoutes(router *mux.Router, sqlDB *sql.DB, wsHandler *wshandler.Handler, logsDir string) {
	debug.Debug("Setting up diagnostics routes")

	if wsHandler == nil {
		debug.Error("WebSocket handler is nil, cannot setup diagnostics routes")
		return
	}

	// Create database wrapper for middleware
	database := &db.DB{DB: sqlDB}

	// Create the diagnostic service
	diagnosticService := diagnostic.NewDiagnosticService(sqlDB, wsHandler, logsDir)

	// Create the handler
	diagnosticHandler := admindiagnostics.NewDiagnosticHandler(diagnosticService, wsHandler)

	// Create admin diagnostics subrouter (requires admin authentication)
	// Note: We use /api/admin/diagnostics because SetupDiagnosticsRoutes is called with root router
	// but frontend expects /api/admin/... paths like other admin routes
	diagnosticsRouter := router.PathPrefix("/api/admin/diagnostics").Subrouter()
	diagnosticsRouter.Use(middleware.RequireAuth(database))
	diagnosticsRouter.Use(middleware.AdminOnly)

	// Download full diagnostic package
	diagnosticsRouter.HandleFunc("/download", diagnosticHandler.DownloadDiagnostics).Methods(http.MethodGet, http.MethodOptions)

	// System info
	diagnosticsRouter.HandleFunc("/system-info", diagnosticHandler.GetSystemInfo).Methods(http.MethodGet, http.MethodOptions)

	// Agent debug status endpoints
	diagnosticsRouter.HandleFunc("/agents", diagnosticHandler.GetAgentDebugStatuses).Methods(http.MethodGet, http.MethodOptions)
	diagnosticsRouter.HandleFunc("/agents/{id:[0-9]+}", diagnosticHandler.GetAgentDebugStatus).Methods(http.MethodGet, http.MethodOptions)

	// Toggle debug mode for specific agent
	diagnosticsRouter.HandleFunc("/agents/{id:[0-9]+}/debug", diagnosticHandler.ToggleAgentDebug).Methods(http.MethodPost, http.MethodOptions)

	// Request/purge logs for specific agent
	diagnosticsRouter.HandleFunc("/agents/{id:[0-9]+}/logs", diagnosticHandler.RequestAgentLogs).Methods(http.MethodGet, http.MethodOptions)
	diagnosticsRouter.HandleFunc("/agents/{id:[0-9]+}/logs", diagnosticHandler.PurgeAgentLogs).Methods(http.MethodDelete, http.MethodOptions)

	// Toggle debug mode for all agents
	diagnosticsRouter.HandleFunc("/agents/debug", diagnosticHandler.ToggleAllAgentsDebug).Methods(http.MethodPost, http.MethodOptions)

	// Server/backend debug status and toggle
	diagnosticsRouter.HandleFunc("/server/debug", diagnosticHandler.GetServerDebugStatus).Methods(http.MethodGet, http.MethodOptions)
	diagnosticsRouter.HandleFunc("/server/debug", diagnosticHandler.ToggleServerDebug).Methods(http.MethodPost, http.MethodOptions)

	// Server log stats and purge
	diagnosticsRouter.HandleFunc("/logs/stats", diagnosticHandler.GetLogStats).Methods(http.MethodGet, http.MethodOptions)
	diagnosticsRouter.HandleFunc("/logs/{directory}", diagnosticHandler.PurgeServerLogs).Methods(http.MethodDelete, http.MethodOptions)

	// PostgreSQL logs check
	diagnosticsRouter.HandleFunc("/postgres-logs-exist", diagnosticHandler.CheckPostgresLogsExist).Methods(http.MethodGet, http.MethodOptions)

	// Nginx logs check
	diagnosticsRouter.HandleFunc("/nginx-logs-exist", diagnosticHandler.CheckNginxLogsExist).Methods(http.MethodGet, http.MethodOptions)

	// Nginx hot-reload
	diagnosticsRouter.HandleFunc("/nginx/reload", diagnosticHandler.ReloadNginx).Methods(http.MethodPost, http.MethodOptions)

	debug.Info("Configured admin diagnostics routes: /api/admin/diagnostics/*")
}
